File: core/src/com/jediterm/terminal/model/TerminalLineIntervalHighlighting.java
package com.jediterm.terminal.model;

import com.jediterm.terminal.TerminalColor;
import com.jediterm.terminal.TextStyle;
import org.jetbrains.annotations.NotNull;

public abstract class TerminalLineIntervalHighlighting {
  private final TerminalLine myLine;
  private final int myStartOffset;
  private final int myEndOffset;
  private final TextStyle myStyle;
  private boolean myDisposed = false;

  TerminalLineIntervalHighlighting(@NotNull TerminalLine line, int startOffset, int length, @NotNull TextStyle style) {
    if (startOffset < 0) {
      throw new IllegalArgumentException("Negative startOffset: " + startOffset);
    }
    if (length < 0) {
      throw new IllegalArgumentException("Negative length: " + length);
    }
    myLine = line;
    myStartOffset = startOffset;
    myEndOffset = startOffset + length;
    myStyle = style;
  }

  public @NotNull TerminalLine getLine() {
    return myLine;
  }

  public int getStartOffset() {
    return myStartOffset;
  }

  public int getEndOffset() {
    return myEndOffset;
  }

  public int getLength() {
    return myEndOffset - myStartOffset;
  }

  public boolean isDisposed() {
    return myDisposed;
  }

  public final void dispose() {
    doDispose();
    myDisposed = true;
  }

  protected abstract void doDispose();

  public boolean intersectsWith(int otherStartOffset, int otherEndOffset) {
    return !(myEndOffset <= otherStartOffset || otherEndOffset <= myStartOffset);
  }

  public @NotNull TextStyle mergeWith(@NotNull TextStyle style) {
    TerminalColor foreground = myStyle.getForeground();
    if (foreground == null) {
      foreground = style.getForeground();
    }
    TerminalColor background = myStyle.getBackground();
    if (background == null) {
      background = style.getBackground();
    }
    return new TextStyle(foreground, background);
  }

  @Override
  public String toString() {
    return "startOffset=" + myStartOffset +
      ", endOffset=" + myEndOffset +
      ", disposed=" + myDisposed;
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/model/TerminalLineIntervalHighlighting.kt
package com.jediterm.terminal.model

import com.jediterm.terminal.TextStyle

abstract class TerminalLineIntervalHighlighting internal constructor(
    line: TerminalLine,
    startOffset: Int,
    length: Int,
    style: TextStyle
) {
    val line: TerminalLine
    val startOffset: Int
    val endOffset: Int
    private val myStyle: TextStyle
    var isDisposed: Boolean = false
        private set

    init {
        require(startOffset >= 0) { "Negative startOffset: " + startOffset }
        require(length >= 0) { "Negative length: " + length }
        this.line = line
        this.startOffset = startOffset
        this.endOffset = startOffset + length
        myStyle = style
    }

    val length: Int
        get() = this.endOffset - this.startOffset

    fun dispose() {
        doDispose()
        this.isDisposed = true
    }

    protected abstract fun doDispose()

    fun intersectsWith(otherStartOffset: Int, otherEndOffset: Int): Boolean {
        return !(this.endOffset <= otherStartOffset || otherEndOffset <= this.startOffset)
    }

    fun mergeWith(style: TextStyle): TextStyle {
        var foreground = myStyle.foreground
        if (foreground == null) {
            foreground = style.foreground
        }
        var background = myStyle.background
        if (background == null) {
            background = style.background
        }
        return TextStyle(foreground, background)
    }

    override fun toString(): String {
        return "startOffset=" + this.startOffset +
                ", endOffset=" + this.endOffset +
                ", disposed=" + this.isDisposed
    }
}
